Add PADDING frame to fill up the datagram with Initial packet

junhochoi opened the pull request:
For Initial packets we need to make the UDP datagram at least 1200
bytes following the transport spec. However when the last packet
in the datagram is Short and the datagram is not fully filled, we try
to do null padding at the end which causes a decryption failure on
the server because Short packet doesn't have an explicit length.

To avoid this issue, when Short packet is added in the same
datagram with Initial packet, add a PADDING frame to Short packet
to fill up the datagram size. If Short packet is not added,
we still add zero padding at the end of packet.

ghedo commented on src/lib.rs:
```diff
@@ -2964,6 +2976,17 @@ impl Connection {
+        // Create PADDING frame for the rest of the packet.
+        // PADDING frame can't be a single frame in the packet,
+        // so only add it when there is a frame already.
+        if add_padding && left >= 1 && !frames.is_empty() {
```
Should probably move this a few lines down after the existing `frames.is_empty()` check, so we don't have to duplicate it.

junhochoi replied:
Moved.

ghedo commented on src/lib.rs:
```diff
@@ -2286,14 +2287,23 @@ impl Connection {
+            }
+
+            let (ty, written) = match self
+                .send_single(&mut out[done..done + left], add_padding)
```
It might be better to simply pass the `has_initial` value to `send_single()` so that we can unify the logic in one place inside it, rather than split in two parts, something like:
```rust
if has_initial && pkt_type == packet::Type::Short && left > 1 ...
```
this also means we wouldn't have to duplicate the `write_pkt_type()` call in `send()`, which seems somewhat "cleaner".

junhochoi replied:
Updated.

ghedo reviewed (changes requested):
Added a couple of comments, but looks good overall:

Should also add a test though. The following seems to work:
```rust
    #[test]
    fn coalesce_padding_short() {
        let mut buf = [0; 65535];

        let mut pipe = testing::Pipe::default().unwrap();

        // Client sends first flight.
        let len = pipe.client.send(&mut buf).unwrap();
        assert_eq!(len, MIN_CLIENT_INITIAL_LEN);
        assert_eq!(pipe.server.recv(&mut buf[..len]), Ok(len));

        // Server sends first flight.
        let len = pipe.server.send(&mut buf).unwrap();
        assert_eq!(len, MIN_CLIENT_INITIAL_LEN);
        assert_eq!(pipe.client.recv(&mut buf[..len]), Ok(len));

        let len = pipe.server.send(&mut buf).unwrap();
        assert_eq!(pipe.client.recv(&mut buf[..len]), Ok(len));

        // Client sends stream data.
        assert_eq!(pipe.client.is_established(), true);
        assert_eq!(pipe.client.stream_send(4, b"hello", true), Ok(5));

        // Client sends second flight.
        let len = pipe.client.send(&mut buf).unwrap();
        assert_eq!(len, MIN_CLIENT_INITIAL_LEN);
        assert_eq!(pipe.server.recv(&mut buf[..len]), Ok(len));

        // None of the sent packets should have been dropped.
        assert_eq!(pipe.client.sent_count, pipe.server.recv_count);
        assert_eq!(pipe.server.sent_count, pipe.client.recv_count);
    }
```

junhochoi commented:
I added the test you suggested and updated based on your feedback. Thanks!

junhochoi commented:
> Added a couple of comments, but looks good overall:
> 
> Should also add a test though. The following seems to work:
> 
> ```rust
>     #[test]
>     fn coalesce_padding_short() {
>         let mut buf = [0; 65535];
> 
>         let mut pipe = testing::Pipe::default().unwrap();
> 
>         // Client sends first flight.
>         let len = pipe.client.send(&mut buf).unwrap();
>         assert_eq!(len, MIN_CLIENT_INITIAL_LEN);
>         assert_eq!(pipe.server.recv(&mut buf[..len]), Ok(len));
> 
>         // Server sends first flight.
>         let len = pipe.server.send(&mut buf).unwrap();
>         assert_eq!(len, MIN_CLIENT_INITIAL_LEN);
>         assert_eq!(pipe.client.recv(&mut buf[..len]), Ok(len));
> 
>         let len = pipe.server.send(&mut buf).unwrap();
>         assert_eq!(pipe.client.recv(&mut buf[..len]), Ok(len));
> 
>         // Client sends stream data.
>         assert_eq!(pipe.client.is_established(), true);
>         assert_eq!(pipe.client.stream_send(4, b"hello", true), Ok(5));
> 
>         // Client sends second flight.
>         let len = pipe.client.send(&mut buf).unwrap();
>         assert_eq!(len, MIN_CLIENT_INITIAL_LEN);
>         assert_eq!(pipe.server.recv(&mut buf[..len]), Ok(len));
> 
>         // None of the sent packets should have been dropped.
>         assert_eq!(pipe.client.sent_count, pipe.server.recv_count);
>         assert_eq!(pipe.server.sent_count, pipe.client.recv_count);
>     }
> ```

Added this test and updated based on your feedback. Thanks!

ghedo commented:
Thanks, I changed the comment in the code to add more details, but lgtm otherwise, so merged.
